### PR TITLE
Adds cluster name to UI. Closes #1353

### LINKED ISSUE #1353: Use cluster name rather than cluster URL in the UI
When users have IP addresses instead of hostnames for their K8S cluster, it is confusing and hard to find the cluster.

### PR DESCRIPTION
**Pull request by alexec**
A couple of notes about this:

* Cluster may or may not have a name, if the name is absent we name it "in-cluster".
* For short-form name we show the cluster name and have a tooltip for the IP.
* For long-form we should the cluster name followed by the IP (in brackets).

### CODE REVIEW

File: ui/src/app/settings/components/clusters-list/clusters-list.tsx [hunk 1]
```diff
@@ -29,7 +25,7 @@ export const ClustersList = () => (
                             <div className='argo-table-list__row' key={cluster.server}>
                                 <div className='row'>
                                     <div className='columns small-3'>
-                                        <i className='icon argo-icon-hosts'/> {cluster.name}
+                                        <i className='icon argo-icon-hosts'/> {cluster.name || 'in-cluster'}
```
Comment (alexec) [inline ui/src/app/settings/components/clusters-list/clusters-list.tsx]: fix

File: ui/src/app/settings/components/project-edit-panel/project-edit-panel.tsx [hunk 2]
```diff
@@ -62,21 +63,25 @@ export const ProjectEditPanel = (props: {
                     )}
                 </DataLoader>
 
-                <DataLoader load={() => services.clusters.list().then((clusters) => clusters.concat({server: '*'} as models.Cluster).map((cluster) => cluster.server))}>
```
Comment (alexec) [inline ui/src/app/settings/components/project-edit-panel/project-edit-panel.tsx]: this code that adds "*" is moved down

File: ui/src/app/shared/components/cluster.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    const text = showUrl ? clusterTitle(cluster) : clusterName(cluster.name);
+    return <Tooltip content={cluster.server}><span>{text}</span></Tooltip>;
+};
+
+interface Props {
+    url: string;
+    showUrl?: boolean;
+}
+
+export const Cluster = React.memo((props: Props) => {
```
Comment (alexec) [inline ui/src/app/shared/components/cluster.tsx]: we need to cache this look up, it's not extortionate, but could happen 100s of times if we're not careful

Comment (alexec) [reply ui/src/app/shared/components/cluster.tsx]: this cache does not work -- yet

File: ui/src/app/applications/components/application-create-panel/application-create-panel.tsx [hunk 4]
```diff
@@ -6,6 +6,7 @@ import { FieldApi, Form, FormApi, FormField as ReactFormField, Text } from 'reac
 const jsonMergePatch = require('json-merge-patch');
 
 import { AutocompleteField, CheckboxField, TagsInputField, YamlEditor } from '../../../shared/components';
```
Comment (alexmt) [inline ui/src/app/applications/components/application-create-panel/application-create-panel.tsx]: Please export clusterTitle from components/index and change import to 
```
import { AutocompleteField, CheckboxField, TagsInputField, YamlEditor, clusterTitle } from '../../../shared/components';
```

File: ui/src/app/shared/components/cluster.tsx (hunk 3 above)
Comment (alexmt) [reply ui/src/app/shared/components/cluster.tsx]: I guess we need caching on app list page, right? I would load all clusters on list page and pass it via context. The Cluster can check if context has list of cluster and load cluster info via api if context is empty.

Comment (alexec) [reply ui/src/app/shared/components/cluster.tsx]: yes - I don't know how to do this yet - React.memo does not do what I wanted

Comment (alexmt) [reply ui/src/app/shared/components/cluster.tsx]: working on PR to your fork

**Comment by alexec** [comment]
hi @alexmt , I've removed the changes that were a potential performance problem - maybe a second PR could address this - what do you think ?

**Review by alexmt — APPROVED** [review]
Please fix this comment before merging: https://github.com/argoproj/argo-cd/pull/1814/files#r298326659

**Pull request merged**